atlasdb-cassandra/src/test/java/com/palantir/atlasdb/cassandra/ReloadingCqlClusterContainerTest.java

```diff
@@ -0,0 +1,119 @@
+import com.palantir.atlasdb.cassandra.ReloadingCqlClusterContainer.CqlClusterFactory;
+import com.palantir.atlasdb.cassandra.backup.CqlCluster;
+import com.palantir.atlasdb.keyvalue.cassandra.async.client.creation.ClusterFactory.CassandraClusterConfig;
+import com.palantir.atlasdb.timelock.api.Namespace;
+import com.palantir.refreshable.Refreshable;
+import com.palantir.refreshable.SettableRefreshable;
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+@RunWith(MockitoJUnitRunner.class)
+public class ReloadingCqlClusterContainerTest {
+    private static final Namespace NAMESPACE = Namespace.of("TEST");
+    private static final CassandraServersConfig EMPTY_SERVERS_CONFIG = ImmutableDefaultConfig.of();
+    private static final CassandraServersConfig SERVERS_CONFIG = ImmutableDefaultConfig.builder()
+            .addThriftHosts(InetSocketAddress.createUnresolved("foo", 42))
+            .build();
+
+    @Mock
+    private CassandraClusterConfig cassandraClusterConfig;
+
+    @Mock
+    private CqlClusterFactory cqlClusterFactory;
+
+    @Test
+    public void lastCqlClusterClosedAfterClose() throws IOException {
+        when(cqlClusterFactory.create(cassandraClusterConfig, EMPTY_SERVERS_CONFIG, NAMESPACE))
+                .thenReturn(mock(CqlCluster.class));
+
+        Refreshable<CassandraServersConfig> cassandraServersConfigRefreshable = Refreshable.only(EMPTY_SERVERS_CONFIG);
+        ReloadingCqlClusterContainer reloadingCqlClusterContainer = ReloadingCqlClusterContainer.of(
+                cassandraClusterConfig, cassandraServersConfigRefreshable, NAMESPACE, cqlClusterFactory);
+
+        CqlCluster cqlCluster = reloadingCqlClusterContainer.get();
+        reloadingCqlClusterContainer.close();
+        verify(cqlCluster).close();
+    }
+
+    @Test
+    public void previousCqlClusterIsClosedAfterRefresh() throws IOException {
+        CqlCluster firstMock = mock(CqlCluster.class);
+        CqlCluster secondMock = mock(CqlCluster.class);
+        when(cqlClusterFactory.create(cassandraClusterConfig, EMPTY_SERVERS_CONFIG, NAMESPACE))
+                .thenReturn(firstMock);
+        when(cqlClusterFactory.create(cassandraClusterConfig, SERVERS_CONFIG, NAMESPACE))
+                .thenReturn(secondMock);
+
+        SettableRefreshable<CassandraServersConfig> cassandraServersConfigRefreshable =
+                Refreshable.create(EMPTY_SERVERS_CONFIG);
+        ReloadingCqlClusterContainer reloadingCqlClusterContainer = ReloadingCqlClusterContainer.of(
+                cassandraClusterConfig, cassandraServersConfigRefreshable, NAMESPACE, cqlClusterFactory);
+        CqlCluster cqlCluster = reloadingCqlClusterContainer.get();
+        cassandraServersConfigRefreshable.update(SERVERS_CONFIG);
+        verify(cqlCluster).close();
```

Comment:
it's worth verifying that the second cluster is not closed

Reply:
and also that when you call get a second time that you get the second cluster, and not the first

Reply:
ah, it looks like you do that verification in a different test

Reply:
Yeah, decided against testing those things in this test, since they're tested elsewhere.